nano/secure/ledger.cpp

```diff
@@ -1044,18 +1208,53 @@ bool nano::ledger::can_vote (nano::transaction const & transaction_a, nano::bloc
 			if (block != nullptr)
 			{
 				nano::confirmation_height_info height;
-				auto error = store.confirmation_height_get (transaction_a, block->account ().is_zero () ? block->sideband ().account : block->account (), height);
+				auto error = store.confirmation_height_get (transaction_a, block->account (), height);
 				debug_assert (!error);
-				result = block->sideband ().height <= height.height;
+				result = block->height () <= height.height;
 			}
 		}
 		return result;
 	});
 }
 
-bool nano::ledger::is_epoch_link (nano::link const & link_a)
+/*
+ * An epoch link will be the account for self-signed epochs, otherwise will be pre-determined epoch links
+ */
+bool nano::ledger::has_epoch_link (nano::block const & block_a) const
 {
-	return network_params.ledger.epochs.is_epoch_link (link_a);
+	if (block_a.type () >= nano::block_type::state2)
+	{
+		auto state_block = dynamic_cast<nano::state_block const *> (&block_a);
+		if (state_block)
+		{
+			if (state_block->hashables.flags.is_epoch_signer () && network_params.ledger.epochs.is_epoch_link (block_a.link ()))
+			{
+				// Epoch-signed epoch block
+				return true;
+			}
+			else
+			{
+				// Can be a self-signed epoch block
+				return (state_block->hashables.flags.link_interpretation () == nano::link_flag::noop && block_a.link () == block_a.account ());
+			}
+		}
+	}
+
+	return network_params.ledger.epochs.is_epoch_link (block_a.link ());
+}
+
+bool nano::ledger::is_self_signed_epoch (nano::block const & block_a) const
+{
+	if (block_a.type () >= nano::block_type::state2)
+	{
+		auto state_block = dynamic_cast<nano::state_block const *> (&block_a);
+		if (state_block)
+		{
+			return (state_block->hashables.flags.link_interpretation () == nano::link_flag::noop && block_a.link () == block_a.account ());
+		}
+	}
+
+	return false;
 }
```

Comment:
This PR nicely fixes some dynamic_cast's in ipc::flatbuffers_builder, and it would be good to avoid it here as well.

What do you think about making `has_epoch_link` and `is_self_signed_epoch` polymorphic `block` functions? The default implementation would return the non-state2 case, with state_block overriding it.

The reason I'm asking is that `dynamic_cast` has shown up in profiling on some compilers in the past (like #831). I imagine these functions will be called frequently.

It can be argued these doesn't make sense for non-state blocks, but the ledger functions are called on all block types anyway, so I think that's fine (and possibly the lesser code smell vs dynamic_cast). Maybe rename `is_self_signed_epoch` to `has_self_signed_epoch` if the move is done.

In addition to improved performance, this should lead to a tad less code, and there's one less branch for checking the dynamic_cast. For `has_epoch_link`, network_params can be passed in as an argument, as having it as a static local adds synchronization overhead.

Reply:
I guess they could have been `static_cast` too, but your idea of making them members of blocks helps with avoiding downcasts, updated.